lib/jsonapi/utils/underscore.ex

```diff
@@ -25,6 +25,31 @@ defmodule JSONAPI.Utils.Underscore do
     String.replace(value, ~r/([a-zA-Z0-9])-([a-zA-Z0-9])/, "\\1_\\2")
   end
 
+  @doc ~S"""
+  Takes a Map with dasherized keys and makes the underscored.
+
+  ## Examples
+
+      iex> dasherize_keys(%{"foo-bar" => "baz"})
+      %{"foo_bar" => "baz"}
+
+      iex> dasherize_keys(%{"f-b" => %{"a-d" => "z"}, "c-d" => "e"})
+      %{"f_b" => %{"a_d" => "z"}, "c_d" => "e"}
+  """
+  def dasherize_keys(map) when is_map(map) do
```

Comment:
We don't need the guards on `dasherize_keys/1` and `dash/1` if we only use them for specific values. What do you think about using `dash/1` and pattern matching to do our work?

```elixir  
@doc """
  Replace dashes between words in `value` with underscores

  Ignores dashes that are not between letters/numbers

  ## Examples

      iex> dash("top-posts")
      "top_posts"

      iex> dash("-top-posts")
      "-top_posts"

      iex> dash("-top--posts-")
      "-top--posts-"

      iex> dash(%{"foo-bar" => "baz"})
      %{"foo_bar" => "baz"}

      iex> dash({"foo-bar", "baz"})
      {"foo_bar", "baz"}
  """
  def dash(value) when is_binary(value) do
    String.replace(value, ~r/([a-zA-Z0-9])-([a-zA-Z0-9])/, "\\1_\\2")
  end

  def dash(map) when is_map(map) do
    Enum.reduce(map, %{}, &dash/1)
  end

  def dash({key, value}) do
    {dash(key), dash(value)}
  end

  def dash(value) do
    value
  end
```

Reply:
We could also handle atoms though this is really not something we should encourage:

```elixir
def dash(value) when is_atom(value) do
  value
  |> to_string()
  |> dash()
  |> String.to_atom()
end
```

Reply:
Yeah, I'm into that! I'll make that change.

Reply:
I'm 100% into renaming these functions. This was a source of confusion for me for sure!

Reply:
@snewcomer is part of the intention of #137 to rename these methods? If not I'll open a PR to do so.